### PR TITLE
Add order support to API package

### LINKED ISSUE #27873: Add Order Type functionality to `@woocommerce/api` package
Add Order Type functionality to `@woocommerce/api` package.

Order Type:
- Listing operation
- Creation of order model

Tasks:
- [ ] create an empty order
- [ ] Update `/tests/e2e/core-tests/specs/merchant/wp-admin-order-status-filters.test.js` to use the API to create the orders for the test
- [ ] transformer for simple properties
- [ ] transformers for complex properties
- [ ] add read, list, update, delete functionality

### PR DESCRIPTION
**Pull request by zhongruige**
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

This PR completes work on adding order support to the API Package. Additionally, it adds a test around that work and removes an unused import from an existing e2e tests.

Closes https://github.com/woocommerce/woocommerce/issues/27873

### How to test the changes in this Pull Request:

1. Verify the CI runs and passes
2. Run the new test for the API changes and verify it passes `npx wc-e2e test:e2e tests/e2e/specs/rest-api/order.js`
3. Verify the e2e test suite passes

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

### Changelog entry

> Enter a summary of all changes on this Pull Request. This will appear in the changelog if accepted.

### CODE REVIEW

**Comment by rrennick** [comment]
@zhongruige Thanks for the quick turn around on finishing this off. I tested it locally and it works great.

```
Test Suites: 38 passed, 38 total
Tests:       3 skipped, 162 passed, 165 total
Snapshots:   0 total
Time:        851.639s
```

Because I did the models for this I'm going to get @lsinger to review.

File: tests/e2e/api/src/models/orders/shared/types.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
+/**
+ * An order's status.
+ *
+ * @typedef OrderStatus
+ */
+export type OrderStatus = 'pending' | 'processing' | 'complete' | 'on-hold' | 'refunded'
+	| 'cancelled' | 'failed' | 'trash' | string;
```
Comment (lsinger) [inline tests/e2e/api/src/models/orders/shared/types.ts:7]: Unclear why this type ends in `| string` while the others below don't.

File: tests/e2e/api/src/models/orders/shared/classes.ts [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+	 *
+	 * @type {string}
+	 */
+	public readonly subtotal: string = '';
+}
+
+/**
+ * An order address.
+ */
+export class OrderAddress extends Model {
```
Comment (lsinger) [inline tests/e2e/api/src/models/orders/shared/classes.ts:39]: Just thinking out loud: maybe there's an address type that would be useful in other, non-order areas as well, and the OrderAddress would then be a further specialization of that one.

File: tests/e2e/api/src/models/orders/shared/classes.ts [hunk 3]
```diff
@@ -0,0 +1,412 @@
+import { MetaData } from '../../shared-types';
+import { Model } from '../../model';
+import { TaxStatus } from './types';
+
+/**
```
Comment (lsinger) [inline tests/e2e/api/src/models/orders/shared/classes.ts:5]: Thinking out loud: taxes and the associated calculations (e.g., rounding) are important but error-prone. The model though is just for interacting with the API though, which does all the calculations, which takes off some of the pressure. Are we still confident that we're testing this part enough, or do we need some unit tests here? (Possibly I'll discover them further down the file list, not sure.)

File: tests/e2e/api/src/repositories/rest/coupons/transformer.ts [hunk 4]
```diff
@@ -57,6 +57,7 @@ export function createCouponTransformer(): ModelTransformer< Coupon > {
 					maximumAmount: 'maximum_amount',
 					emailRestrictions: 'email_restrictions',
 					usedBy: 'used_by',
+					links: '_links',
```
Comment (lsinger) [inline tests/e2e/api/src/repositories/rest/coupons/transformer.ts:60]: Just curious: why the underscore here?

File: tests/e2e/core-tests/specs/api/order.test.js [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+ const {
+	 it,
+	 describe,
+	 beforeAll,
+ } = require( '@jest/globals' );
+
+ /**
+ * Creates an order and tests interactions with it via the API.
+ */
+const runOrderApiTest = () => {
```
Comment (lsinger) [inline tests/e2e/core-tests/specs/api/order.test.js:20]: OK, here we have the test for the order API! 👍 

Given that the fixture for the order (the one in `tests/e2e/config/default.json`) is pretty bare-bones (no address, no line items, therefore no taxes, etc.), I wonder whether we need to add tests for more complicated orders. Probably yes? Not in this PR, but as a follow-up.

**Review by lsinger — APPROVED** [review]
CI runs and passes, all e2e tests and specifically the new order test pass locally. 

I left a few comments, none of which are blockers -- mostly questions to help me learn the codebase. Since this is a slightly larger PR I'll leave it to @zhongruige to merge. 

:shipit:

File: tests/e2e/api/src/models/orders/shared/types.ts (hunk 1 above)
Comment (rrennick) [reply tests/e2e/api/src/models/orders/shared/types.ts:7]: It allows for custom statuses.

File: tests/e2e/api/src/models/orders/shared/classes.ts (hunk 2 above)
Comment (rrennick) [reply tests/e2e/api/src/models/orders/shared/classes.ts:39]: That's possible but we don't have any of the other addresses to use as a model for deciding between base and specialization. So long as the change were backward compatible we can implement the specialization as soon as we have a use case for it.

File: tests/e2e/api/src/models/orders/shared/classes.ts (hunk 3 above)
Comment (rrennick) [reply tests/e2e/api/src/models/orders/shared/classes.ts:5]: Adding additional API tests is on the post `API v0.2` list.

File: tests/e2e/api/src/repositories/rest/coupons/transformer.ts (hunk 4 above)
Comment (rrennick) [reply tests/e2e/api/src/repositories/rest/coupons/transformer.ts:60]: That comes from the rest API. I suspect that it originated either from being a read-only field or a WordPress standard at some point in the past.

File: tests/e2e/core-tests/specs/api/order.test.js (hunk 5 above)
Comment (rrennick) [reply tests/e2e/core-tests/specs/api/order.test.js:20]: Follow up sounds good.

**Comment by lsinger** [comment]
@rrennick thanks for your responses! 🙌

File: tests/e2e/core-tests/specs/api/order.test.js (hunk 5 above)
Comment (zhongruige) [reply tests/e2e/core-tests/specs/api/order.test.js:20]: Created an issue for this here: https://github.com/woocommerce/woocommerce/issues/29900

**Comment by zhongruige** [comment]
Thank you for the review @lsinger!

**Pull request merged**